src/ztex/ztex.c

```diff
@@ -69,6 +71,23 @@ int vendor_request(struct libusb_device_handle *handle, int cmd, int value, int
 //
 ////////////////////////////////////////////////////////////////////////////////
 
+// checks if given string is a valid Serial Number
+int ztex_sn_is_valid(char *sn)
+{
+	int i;
+	for (i = 0; i < ZTEX_SNSTRING_LEN; i++) {
+		if (!sn[i])
+			return i < ZTEX_SNSTRING_MIN_LEN ? 0 : 1;
+		if ( !( (sn[i] >= '0' && sn[i] <= '9') || (sn[i] >= 'A' && sn[i] <= 'F')
+				|| (sn[i] >= 'a' && sn[i] <= 'f')) )
+			return 0;
+	}
+	if (sn[i])
+		return 0;
+
+	return 1;
+}
+
```

Comment:
Is there a risk to run into issues similar to what we had with non-default firmware - someone having a serial number patched to be of non-standard format? Or are they converted from raw binary data to hex by the firmware itself, so have to be of the correct format as long as the firmware is correct? BTW, I noticed a serial number may change from fake but correct format (could it as well not have been correct format?) to real upon our firmware upload - but I guess we won't even proceed with firmware upload if the initially seen SN is not of correct format.

We should be able to use and ideally refer to with the `--devices` option to all devices, even those with incorrect SNs, although I understand this can't be always possible as e.g. two devices might have not only incorrect, but also matching SNs.

Reply:
According to USB spec, SN can be arbitrary string up to 253 bytes. That means a corrupted or intentionally developed firmware can return anything. (In practice Ztex SDK assumes that a descriptor fits into one 64-byte hardware buffer EP0, hence 62 bytes limit).

A user can set any valid SN using FWLoader or API. In such case original SN is preserved and remains hidden, user-defined SN is serviced on USB Get String Descriptor requests. That's if firmware contains standard routines from Ztex SDK. John firmware also behaves this way.

Firmware on one of your boards may be the one that does not honor user-defined SN, uses only original SN.

I've checked include files from Ztex SDK, it operates as follows:
- There're two places where SN is stored: user-modified settings and constant settings. Both data structures reside in external I2C-connected EEPROM chip. On MCU startup, it reads user settings first. 10-byte SN is stored in ASCII and there's no check for valid hex on MCU side in either read or write function. If user-defined SN equals to "0000000000" then it proceeds to reading constant settings (*). There're 5 bytes which are converted to hex.

(*) These constant settings might be called constant only because standard firmware never writes there.

If there's a board with some incorrectly behaving firmware, the user can be advised to:
1. power-cycle the board with JP1 (the only jumper) switched. It boots some "internal firmware";
2. then upload "default firmware" or john's firmware into EEPROM using FWLoader. (this can be skipped).